content/client-lib-development-guide/features.textile

```diff
@@ -232,6 +232,19 @@ h3(#rest-auth). Auth
 ** @(RSA16b)@ If the library is provided with a token without the corresponding @TokenDetails@, then this holds a @TokenDetails@ instance in which only the @token@ attribute is populated with that token string
 ** @(RSA16c)@ Is set with the current token (if applicable) on instantiation and each time it is replaced, whether the result of an explicit @Auth#authorize@ operation, or a library-initiated renewal resulting from expiry or a token error response
 ** @(RSA16d)@ Is empty if there is no current token, including after a previous token has been determined to be invalid or expired, or if the library is using basic auth
+* @(RSA17)@ @Auth#revokeTokens@ function: This function is intended to be used to revoke tokens using provided @TokenRevocationRequest@ details. It internally creates request for  "REST revokeTokens endpoint":/rest-api/#revokeTokens and handles the response.
+** @(RSA17a)@ Method signature is @revokeTokens(TokenRevocationRequest)@.
+** @(RSA17b)@ If the @TokenRevocationRequest#key@ attribute is not provided, then the function must throw an error with a @"Key is not provided"@ message.
+** @(RSA17c)@ If the @TokenRevocationRequest#targets@ attribute is null or empty (as per @TF3@), then the function must throw an error with a @"targets cannot be null or empty"@ message.
+** @(RSA17d)@ If the @TokenRevocationRequest#key@ is provided and @TokenRevocationRequest#targets@ valid, then the function must create the request and execute it.
+** @(RSA17e)@ Function returns a @TokenRevocationResponse@ object that contains the result of the request if request is successful.
+** @(RSA17f)@ Function throws an appropriate error/exception if the request fails either for the reasons of invalid request, or as  a result of request failing.
+** @(RSA17g)@ Tests in this section must exist in client library.
+*** @(RSA17g1)@ Given a null @TokenRevocationRequest#key@ attribute, verify that the function throws an error with a @"Key is not provided"@ message.
+*** @(RSA17g2)@ Given a null or empty @TokenRevocationRequest#targets@ attribute, verify that the function throws an error with a @"targets cannot be null or empty"@ message.
+*** @(RSA17g3)@ Given a non-null @TokenRevocationRequest#key@ attribute and a valid @TokenRevocationRequest#targets@ attribute, verify that the function creates the request and executes it.
+*** @(RSA17g4)@ Given a non-null but invalid @TokenRevocationRequest#key@ attribute and a valid @TokenRevocationRequest#targets@ attribute, verify that client library throws an API error.
+*** @(RSA17g5)@ Given a valid @TokenRevocationRequest#key@ attribute and a valid @TokenRevocationRequest#targets@ attribute, verify that the function executes the request and returns a @TokenRevocationResponse@ with targets matching with the original targets in request.
```

Comment:
Don't think client libs can have this test now, they're just creating simulation apps which won't have revocation enabled

(we might be able to have them specify in the appspec that we want revocable keys etc, but not sure that'll work as-is right now)

Reply:
Removed all tests here 6dafadef8882fc493a11ee695cb4c627ca8b1de5

Reply:
> they're just creating simulation apps which won't have revocation enabled

Why not?  This seems insane that we are creating simulation apps that don't have a feature we need for end-to-end tests in our SDKs?  @SimonWoolf can this not be trivially supported?

In addition, what is the plan for token revocation more broadly?  Is our plan not to make this feature self-servicable?  Why are we holding off offering this to customers?

Reply:
> Why not?

No reason not to, was just an oversight on my part -- done now https://github.com/ably/realtime/pull/4143

> In addition, what is the plan for token revocation more broadly? Is our plan not to make this feature self-servicable? Why are we holding off offering this to customers?

My understanding was it's in beta until we have a few customers using it, who can give feedback to us about how it works in practice, whether the API needs any changes, etc., at which point we'll make any final API tweaks, then enable it (defaulting to 1 revocation shard) for everyone.

I can't find the conversation now, but IIRC I was chatting to @ben-pickering about the status of this a few months ago and he said that there had been slower than expected progress getting customers to try the beta, so this process was taking a bit of time.

Reply:
> defaulting to 1 revocation shard
Would we consider just making the shard count a function of the number of connections in the package? (I was going to suggest the number of `tokenRequests` or `apiRequests`, but with JWT neither of those correlate with actual token usage. We would set the shards as a simple multiple of the peak connections, and only ever increase it from that point forward

Reply:
> Would we consider just making the shard count a function of the number of connections in the package?

Aren't we about to move to a pricing model where we're not using packages/quotas for self-service any more? So we wouldn't have that information.

I was assuming that for self-service we were just going to have it be a fixed limit (1 shard = 50 revocations per second), and for business/enterprise who can request a higher limit, we'd just divide whatever limit they've requested/negotiated by 50 and set that as the number of shards

Reply:
ok, makes sense, thanks